pkg/workspace/apply_test.go

```diff
@@ -522,3 +722,23 @@ func TestApply(t *testing.T) {
 		})
 	}
 }
+
+func TestApplyWithMissingWorkspaceDeclaration(t *testing.T) {
+	names.TestingSeed()
+	ts := v1beta1.TaskSpec{
+		Steps:      []v1beta1.Step{{}},
+		Sidecars:   []v1beta1.Sidecar{{}},
+		Workspaces: []v1beta1.WorkspaceDeclaration{}, // Intentionally missing workspace declaration
+	}
+	bindings := []v1beta1.WorkspaceBinding{{
+		Name: "source",
+		PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+			ClaimName: "testpvc",
+		},
+	}}
+	vols := workspace.CreateVolumes(bindings)
+	_, err := workspace.Apply(ts, bindings, vols)
+	if err == nil || !strings.Contains(err.Error(), "did not exist in declared workspaces") {
```

Comment:
nit: i find that asserting against error messages (or log messages) is a bit brittle to maintain over time - if it's important to assert that the correct error is returned, i usually go to one of 2 alternatives:

* return different types of errors  (https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully is a very interesting post on error handling in Go and iirc doenst like that approach either)
* treat the different error types as a sign that one function is doing too much, i.e. if it can fail in 2 different ways and its important to distinguish between both, maybe you're better off with two separate functions (it does SEEM in this case like maybe Apply is also doing validation)

Reply:
Removed string error message check.